### ISSUE TITLE
Request for Guidance on Reproducing MoR

### ISSUE THREAD

**Issue by franksfc** [raymin0223/mixture_of_recursions#15]
Dear authors,

First of all, thank you for releasing Mixture-of-Recursions (MoR) and the training code!
I’m an undergraduate student very eager to learn efficient transformer design and recursive computation.
I’m currently trying to reproduce the 135M-scale MoR model (the one listed in Table 3, Expert-Choice Router, Recursion = 3, Middle-Cycle sharing, recursion-wise KV cache).

I followed the paper’s setup closely. Here is my config:
name: 
wandb: true
wandb_mode: online
wandb_entity:
wandb_project: mor
wandb_run_name: null
output_dir: null
wandb_run_id: 
tensorboard: false
tensorboard_dir: null
resume_from_checkpoint: false
resume_step: null
dataset: fineweb_edu
weights: null
total_batch_size: 1024
per_device_train_batch_size: 16
gradient_accumulation_steps: null
batch_size_rampup_steps: null
max_length: 2048
add_bos_token: false
global_shuffling: false
local_shuffling: false
tokenizer: smollm
model: smollm
model_name_or_path: HuggingFaceTB/SmolLM-135M
model_config: null
attn_implementation: flash_attention_2
use_pretrained_weights: false

recursive:
  enable: true
  base_depth: null
  num_recursion: 3
  sharing: middle_cycle
  ln_share: true
  initialization: random

kv_sharing:
  enable: false
  base_depth: null
  num_recursion: null
  sharing: null
  update_cache: false

relaxation:
  enable: true
  skip_first_loop: false
  method: lora
  lora:
    r: 16
    alpha: 32
    dropout: 0.1
    target_modules:
    - q_proj
    - v_proj
    rank_pattern: null
    alpha_pattern: 2.0
    svd_init: false
  prompt:
    len: 16

mor:
  enable: true
  type: expert
  capacity: "1.0,0.67,0.33"
  rand_router: false
  router_type: linear
  z_loss: false
  z_coeff: 1.0e-05
  temp: 1.0
  expert:
    cap_warmup_step: 1500
    router_func: sigmoid
    alpha: 0.1
    sampling: aux_loss
    include_first: true
    coeff: 0.001
    gating: weighted
  token:
    bal_warmup_step: 0
    router_func: softmax
    alpha: 1.0
    balancing: loss
    coeff: 0.1
    u: 0.001
    gating: weighted

lr_scheduler_type: warmup_stable_decay
lr_scheduler_kwargs:
  num_decay_steps: 0.9
  decay_type: linear
learning_rate: 0.003
adam_beta1: 0.9
adam_beta2: 0.95
weight_decay: 0.1
precision: bf16
max_grad_norm: 1.0
num_train_steps: 14305
stop_steps: 14305
num_warmup_steps: 715
save_interval: 0.25
save_steps: null
fixed_save_steps: null
save_total_limit: null
logging_steps: 1
dataloader_num_workers: 0
gradient_checkpointing: false
deepspeed: ds_configs/stage2.config

evaluation:
  enable: false
  eval_steps: 5
  batch_size: 16
  tasks: lambada_openai,hellaswag,piqa,winogrande,arc_easy,arc_challenge,openbookqa,mmlu,mmlu_continuation,truthfulqa
  device: null
  num_fewshot: null

<img width="1699" height="1187" alt="Image" src="https://github.com/user-attachments/assets/76e1a79e-1dad-420d-a4cb-80d595613b0f" />

After ≈14 k steps (≈ 16.5 e18 FLOPs), my validation loss remains ≈ 3.2, which is significantly higher than the reported ≈ 2.8 in Table 3.
Could you please advise what might account for the difference? I would greatly appreciate any guidance or hints to better match the paper’s result.
Additionally, could you please share (if possible) the exact hyperparameter settings used for the Table 3 MoR (Expert, Recursion = 3, Middle-Cycle sharing, KV-cache, 16.5flops) run? Having these concrete details would be immensely helpful for faithfully reproducing the reported results and for understanding the practical training dynamics of recursive mixtures.

Once again, thank you very much for your time and for open-sourcing such an elegant work!

Best regards,
Frank Song
Undergraduate Student

**Comment by kimyuji** [comment]
Hi, Frank! 
Thank you for your interest in our paper! 

First, Table 3 in our paper shows the results for the 360M base model. For the 135M base model results, refer to Table 9. 
Based on that, the validation loss you're looking for (MoR with recursion 3, based on smollm-135M and trained with 16.5EB FLOPs) should be 2.949.

We went through your config and found a few corrections needed to reproduce the result:
 
1. The base model's number of layers should be modified (set num_hidden_layers: 32).
2. Relaxation is not used in our MoR setup.
3. We did not use capacity warmup (set cap_warmup_step: 0).
4. To match 16.5EB FLOPs, the model should be trained much longer(36326 steps).
5. The num_decay_steps and num_warmup_steps also need to be set correctly (as shown in the config below).

Here is the part of the configuration we used to get the result for MoR-3 with the 135M base model (16.5EB FLOPs):

```
total_batch_size: 1024
per_device_train_batch_size: 16
gradient_accumulation_steps: null
batch_size_rampup_steps: null
max_length: 2048

tokenizer: smollm
model: smollm
model_name_or_path: HuggingFaceTB/SmolLM-135M
model_config: 
  num_hidden_layers: 32
attn_implementation: flash_attention_2
use_pretrained_weights: false

recursive:
  enable: true
  base_depth: null
  num_recursion: 3
  sharing: middle_cycle
  ln_share: true
  initialization: random

kv_sharing:
  enable: false

relaxation:
  enable: false

mor:
  enable: true
  type: expert
  capacity: null
  rand_router: false
  router_type: linear
  z_loss: false
  z_coeff: 0.001
  temp: 1.0

  expert:
    cap_warmup_step: 0
    router_func: sigmoid
    alpha: 0.1
    sampling: aux_loss
    include_first: true
    coeff: 0.001
    gating: weighted

lr_scheduler_type: warmup_stable_decay
lr_scheduler_kwargs:
  num_decay_steps: 7265
  decay_type: linear
learning_rate: 0.003
adam_beta1: 0.9
adam_beta2: 0.95
weight_decay: 0.1
precision: bf16
max_grad_norm: 1.0
num_train_steps: 36326
stop_steps: 36326
num_warmup_steps: 220
save_interval: 0.1
save_steps: null
fixed_save_steps: '3523,8806,29061'
save_total_limit: null
logging_steps: 100
dataloader_num_workers: 0
gradient_checkpointing: false
deepspeed: ds_configs/stage2.config
```

We hope this helps, and please let us know if you have any other questions!

Best,
Yujin Kim

**Comment by raymin0223** [comment]
@franksfc 

Please feel free to re-open the issue if you still have the problem to reproduce the results.
As @kimyuji pointed, result in Table 3 actually used MoR-360m model checkpoint, not 135m checkpoint!

You have to use `SmolLM-360M` and then apply `num_recursion=3` to make 118m non-embedding parameter sized model.
